FHEM/14_SD_WS.pm

```diff
@@ -1539,20 +1539,22 @@ sub SD_WS_Parse {
         #        Nibble: 01 23 45 67 89 01 23 45 67 89 01 23 45 67 
         # aa aa aa 2d d4 40 01 1C DF 8F 00 00 97 62 20 A6 80 28 01       -> ID  0x11cdf
         #                40 01 3E 3C 90 00 00 10 5B A0 2A                -> ID: 0x13e3c
-        #                FF II II II ?? RR RR XX AA ?? ?? ?? ?? ?? ?? ??
+        #                FF II II II ?B RR RR XX AA ?? ?? ?? ?? ?? ?? ??
         # FF:   Family code always 0x40
         # II:   ID (1 byte)
+        #  B:   Voltage of battery / 10 => F = 15 = 1.5v    Not all models have battery reporting. Firest seen in late 2022
         # RR:   the rain bucket tip count => 0.1mm increments
         # XX:   CRC8 of the preceding 5 bytes (Polynomial 0x31, Initial value 0x00, Input not reflected, Result not reflected)
         # SS:   Sum-8 of the preceding 5 bytes 
         # ??:   Unknown Data
-
         sensortype => 'WH40',
         model      => 'SD_WS_126_R',
         prematch   => sub {my ($rawData,undef) = @_; return 1 if ($rawData =~ /^40/); },
         id         => sub {my ($rawData,undef) = @_; return (substr($rawData,2,6));},
         rain_total => sub {my ($rawData,undef) = @_; return 0.1 * hex(substr($rawData,10,4)); },
         rawRainCounter =>  sub {my ($rawData,undef) = @_; return hex(substr($rawData,10,4)); },
+        bat => sub {my ($rawData,undef) = @_; return substr($rawData,9,1) ne '0' ? hex(substr($rawData,9,1)) > 11 ? 'ok' : 'low' : undef; },
+        batVoltage => sub {my ($rawData,undef) = @_; return substr($rawData,9,1) ne '0' ? hex(substr($rawData,9,1)) / 10 : undef; },
```

Comment:
Warum nimmst du das Bit 0 von Nibble 8 nicht mit dazu?
0x90, wie in den ersten Nachrichten von dir würde jetzt als 0 Volt angezeigt. Richtig wäre aber 1,6 Volt.
Im Quellcode von rtl_433 ist das ja auch so:

```
int battery_v  = (b[4] & 0x1f);
"battery_V",        "Battery Voltage", DATA_COND, battery_v != 0, DATA_FORMAT, "%f V", DATA_DOUBLE, battery_v * 0.1f,
```

Reply:
Die Nachricht mit 0x90 ist glaube ich von einer anderen wh40. Ob die Batterie übermittelt weiß ich nicht 

Ich habe mir das so gedacht:
Da kann eine 1,5 V Batterie rein. Also müsste die Skala nur bis 1,5 gehen. 1,6 V ist zu viel.
Mit den vier Bits kam ich exakt auf 1,5 und alle Werte bis 0 können abgebildet werden.

Ob meine Annahme richtig ist weiss ich nicht, aber zwei Messwerte erschienen mir plausibel und wäre eine einfache Implementierung.

Warum im rtl Projekt angenommen wird es könnten auch 5 Bits sein verstehe ich nicht.

Reply:
> Die Nachricht mit 0x90 ist glaube ich von einer anderen wh40. Ob die Batterie übermittelt weiß ich nicht

Mhmm, die Nachrichten https://github.com/RFD-FHEM/RFFHEM/pull/1164#issuecomment-1498179783 sind nicht von deinem WH40?
Das wäre dann schon ziemlich unwahrscheinlich, das dort die gleiche ID verwendet wird.

> Ich habe mir das so gedacht: Da kann eine 1,5 V Batterie rein. Also müsste die Skala nur bis 1,5 gehen. 1,6 V ist zu viel. Mit den vier Bits kam ich exakt auf 1,5 und alle Werte bis 0 können abgebildet werden.

Die meisten neuen Alkaline-Zellen, die ich bisher so gemessen habe, starten mit ca. 1,62 - 1,63 Volt.

> Ob meine Annahme richtig ist weiss ich nicht, aber zwei Messwerte erschienen mir plausibel und wäre eine einfache Implementierung.

Ich sehe bisher von dir 3 Messwerte: 1,6 - 1,5 - 1,3

Reply:
Die 13E3C ist meine.
Hmm vermutlich hast Du Recht, dass es 5 Bits sind

Reply:
@elektron-bbs 

So ich habe den Code nun auf 5 Bit angepasst

Reply:
Ich denke, das passt jetzt.